### PR TITLE
[move-prover] Eliminate the need for weak borrow edges.

### PR DESCRIPTION
**Pull request by wrwg**
This PR refactors the borrow analysis and memory instrumentation to use inter-procedural analysis to determine borrow behavior of called functions. This together with some other measurements eventually allows to eliminate weak edges. The absence of weak edges is essential for the monomorphic translation scheme, as weak edge update is not typable without universal polymorphism in the runtime.

The new approach does the following:

- For each Move function, we create a summary for the borrow graph, which is then consumed by callers. In order to implement summaries, we introduce a new type of edges, so-called *hyper edges*, which summarize a path in the borrow graph. For example, given the following function
  ```
   fun return_ref_different_path_vec2(b: bool, x: &mut V): &mut u64 {
       if (b) Vector::borrow_mut(&mut x.is, 1) else &mut (Vector::borrow_mut(&mut x.ts, 0)).value
   }
   ```
    ... we generate the summary:
    ```
    borrows_from: Return(0) -> {(.is/[], Reference($t1)), (.ts/[]/.value, Reference($t1))}
    ```
   Here, `.is/[]` is a hyper edge with two steps (select field `.is` then do vector index `[]`). The summary says that the return value of this function has two different choices of borrowing, in addition to the one mentioned, `.ts/[]/.value`.

- We also extend the approach how and when to check whether a `write_back` is active or not. The new operation `res := IsParent(BorrowNode, BorrowEdge)(child)` checks whether a child mutation has the given node as an effective parent. This check was previously build into `write_back`; by pulling it out we can emit it more selectively and only when needed. Moreover, the checks implementation had been regressed since the mono backend, and is now fully back with a set of dedicated tests.

- We also change the approach when data invariants are enforced ('pack'). We now uniformly do this exactly when and only when a value is written back to a root location. This leads to a much simpler semantics than previously. Any special casing of private/public functions has been removed.

A new set of tests has been added in `functional/mut_ref.move`. These tests exercise multiple scenarios of functions returning `&mut`.

For the new interpreter, a TODO has been added to implement `IsParent`. Right now it always returns true, which should be fine with the existing sets of tests (and even the Diem framework).

## Motivation

Memory model

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

Existing and new tests.

## Related PRs

NA

### CODE REVIEW

**Comment by shazqadeer** [comment]
A few questions about the PR description:

borrows_from: Return(0) -> {(.is/[], Reference($t1)), (.ts/[]/.value, Reference($t1))}

In the above summary, is $t1 referring to parameter `x` of the function?

What happens if there are two parameters  `x` and `y` and the returned reference could be borrowed (in the same way as now) from either one of them?

**Comment by wrwg** [comment]
> A few questions about the PR description:
> 
> borrows_from: Return(0) -> {(.is/[], Reference($t1)), (.ts/[]/.value, Reference($t1))}
> 
> In the above summary, is $t1 referring to parameter `x` of the function?

Yes parameters are numbered $t0..$tn.

> 
> What happens if there are two parameters `x` and `y` and the returned reference could be borrowed (in the same way as now) from either one of them?

Then you will have hyper edges from both parameters. See end of [this baseline file](https://github.com/diem/diem/blob/2f99f8ec4bf11ee28047e70117c879c16d983ea0/language/move-prover/bytecode/tests/borrow_strong/mut_ref.exp), part of this PR, for summaries of various scenarios.

File: language/move-prover/bytecode/src/borrow_analysis.rs [hunk 1]
```diff
@@ -254,90 +203,170 @@ impl BorrowInfo {
...
                     .entry(dst.clone())
                     .or_default()
-                    .insert(src.clone(), edges.clone());
+                    .insert((src.clone(), edge.clone()));
+            }
+        }
+    }
+
+    /// Collect those leaves which are returned and summarize them in a hyper edge.
+    /// Each of those leaves has a path `out_mut -> ref1 .. -> refn -> in_mut`.
```
Comment (shazqadeer) [inline language/move-prover/bytecode/src/borrow_analysis.rs:224]: Have you swapped in_mut and out_mut in the comment?  The hyper-edge is from the input mutable reference to the output mutable reference as far as I understand.

File: language/move-prover/bytecode/src/borrow_analysis.rs [hunk 2]
```diff
@@ -254,90 +203,170 @@ impl BorrowInfo {
...
+    /// We create a hyper edge `out_mut --summarize(ref1, .., refn)-> in_mut` for it.
+    fn summarize(
```
Comment (shazqadeer) [inline language/move-prover/bytecode/src/borrow_analysis.rs:226]: I don't understand how the summarization is happening.  It seems like something akin to "existential quantification" of all edges incident to or from local variables of the function must be happening.  So, for example, suppose there is an input parameter x, a local variable l, an output parameter y, where all are mutable variables.  Suppose there is an edge e from x to l and another edge f from l to y, then the summary should contain [e,f].  In fact, we have to consider all such combinations to generate all possible vectors of edges.

**Review by shazqadeer — APPROVED** [review]

File: language/move-prover/bytecode/src/stackless_bytecode.rs [hunk 3]
```diff
@@ -256,24 +259,28 @@ impl BorrowNode {
...
+    Hyper(Vec<BorrowEdge>),
 }
 
-/// A borrow edge -- used in memory operations
-#[derive(Eq, PartialEq, Debug, Clone)]
-pub enum BorrowEdge {
-    Weak,
-    Strong(StrongEdge),
+impl BorrowEdge {
+    pub fn flatten(&self) -> Vec<&BorrowEdge> {
```
[the remark this answers is not in the document]
Comment (wrwg) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:276]: Good catch! Yes, this assumption is there, but it was violated in hyper edge construction, which could have hit if `f(..):&mut T` calls `g(..):&mut R`. Now fixed.

File: language/move-prover/bytecode/src/borrow_analysis.rs (hunk 2 above)
Comment (wrwg) [reply language/move-prover/bytecode/src/borrow_analysis.rs:226]: We are construction all paths from in to out. This is happening in function construct_hyper_edges. As you suspect these are then all combinations. So a graph with 2 x 2 branches will end up with 4 hyper edges after summarization.

**Pull request closed without merging**